### PR TITLE
CacheDisk enhancements and bugfixes

### PR DESCRIPTION
**Pull request by isbm**
### What does this PR do?

- Enhances API for all caches (currently CacheDict and CacheDisk)
- Provides `store` function that will trigger persistence manually (in case implementation has one)
- Adds better logging
- "Todoes" TODO hacks
- Guards against IO/OS crashes
- other minor improvements

### What issues does this PR fix or reference?

This all started here as @rongzeng54 spotted a bug, where CacheDisk is not storing pillar data, when its _content_ is changed: https://github.com/saltstack/salt/pull/50179

### New Behavior

- Mentioned bug is fixed
- No crashes on e.g. different permissions (happens when master was previously run as `root`, then as `salt`).
- Everything else should work as before

### Tests written?

- Refactored existing
- Added new

### CODE REVIEW

File: salt/utils/cache.py [hunk 1]
```diff
@@ -130,41 +167,50 @@ def __delitem__(self, key):
...
-            }
-            msgpack.dump(cache, fp_, use_bin_type=True)
+        if msgpack is None:
+            log.error('Cache cannot be read from the disk: msgpack is missing')
+        elif not os.path.exists(self._path):
+            log.debug('Cache path does not exist for reading: %s', self._path)
+        else:
+            try:
+                with salt.utils.files.fopen(self._path, 'rb') as fp_:
+                    cache = salt.utils.data.decode(msgpack.load(fp_, encoding=__salt_system_encoding__))
```
Comment (cachedout) [inline salt/utils/cache.py]: This looks like it will stacktrace if msgpack is missing.

**Review by cachedout — CHANGES_REQUESTED** [review]

File: salt/utils/cache.py (hunk 1 above)
Comment (isbm) [reply salt/utils/cache.py]: Hmm, how? Line #176 will just log error about it and return `None`. Unit tests are also explicitly testing this [here](https://github.com/saltstack/salt/pull/50585/files#diff-9fc11a5086d51e0723e81f3cf065bf0aR171) and implicitly [here](https://github.com/saltstack/salt/pull/50585/files#diff-9fc11a5086d51e0723e81f3cf065bf0aR374) and [here](https://github.com/saltstack/salt/pull/50585/files#diff-9fc11a5086d51e0723e81f3cf065bf0aR362). Am I still missing something?

Comment (brejoc) [reply salt/utils/cache.py]: I think this should be fine. We are only entering else if `msgpack` is not `None`.

**Review by DmitryKuzmenko — APPROVED** [review]

File: salt/utils/cache.py (hunk 1 above)
Comment (cachedout) [reply salt/utils/cache.py]: Wow, what was I thinking there? My fault. Apologies.

**Review by cachedout — APPROVED** [review]

**Pull request merged**